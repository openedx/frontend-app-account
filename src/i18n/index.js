// Placeholder be overridden by `make pull_translations`
export default {
  ar: {},
  'zh-hk': {},
  'zh-cn': {},
  uk: {},
  'tr-tr': {},
  th: {},
  te: {},
  ru: {},
  'pt-pt': {},
  'pt-br': {},
  'it-it': {},
  id: {},
  hi: {},
  he: {},
  'fr-ca': {},
  fa: {},
  'es-es': {},
  'es-419': {},
  el: {},
  'de-de': {},
  da: {},
  bo: {},
};
