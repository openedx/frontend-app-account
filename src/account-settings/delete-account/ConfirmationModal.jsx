import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  AlertModal,
  Button, Form, ActionRow,
} from '@openedx/paragon';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { faExclamationCircle, faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { getConfig } from '@edx/frontend-platform';
import messages from './messages';
import Alert from '../Alert';
import PrintingInstructions from './PrintingInstructions';

export class ConfirmationModal extends Component {
  /**
   * @returns String The message id for a short description of the error, suitable for a header or
   * as the error message under an input field.
   */
  getShortErrorMessageId(reason) {
    switch (reason) {
      case 'empty-password':
        return 'account.settings.delete.account.error.no.password';
      case 'invalid-password':
        return 'account.settings.delete.account.error.invalid.password';
      default:
        return 'account.settings.delete.account.error.unable.to.delete';
    }
  }

  renderError(reason) {
    const { errorType, intl } = this.props;

    if (errorType === null) {
      return null;
    }
    const headerMessageId = this.getShortErrorMessageId(errorType);
    const detailsMessageId = reason === 'empty-password'
      ? null
      : 'account.settings.delete.account.error.unable.to.delete.details';

    return (
      <Alert
        className="alert-danger mt-n2"
        icon={<FontAwesomeIcon className="mr-2" icon={faExclamationCircle} />}
      >
        <h6>{intl.formatMessage(messages[headerMessageId])}</h6>
        {detailsMessageId ? (
          <p className="text-danger">{intl.formatMessage(messages[detailsMessageId])}</p>
        ) : null}
      </Alert>
    );
  }

  render() {
    const {
      status,
      errorType,
      intl,
      onCancel,
      onChange,
      onSubmit,
      password,
    } = this.props;
    const open = ['confirming', 'pending', 'failed'].includes(status);
    const passwordFieldId = 'passwordFieldId';
    const invalidMessage = messages[this.getShortErrorMessageId(errorType)];

    // TODO: We lack a good way of providing custom language for a particular site.  This is a hack
    // to allow edx.org to fulfill its business requirements.
    const deleteAccountModalText2MessageKey = getConfig().SITE_NAME === 'edX'
      ? 'account.settings.delete.account.modal.text.2.edX'
      : 'account.settings.delete.account.modal.text.2';

    return (
      <AlertModal
        isOpen={open}
        title={intl.formatMessage(messages['account.settings.delete.account.modal.header'])}
        onClose={onCancel}
        isOverflowVisible
        footerNode={(
          <ActionRow>
            <Button variant="link" onClick={onCancel}>Cancel</Button>
            <Button variant="danger" onClick={onSubmit}>Yes, Delete</Button>
          </ActionRow>
        )}
      >
        <div className="p-3">
          {this.renderError()}
          <Alert
            className="alert-warning mt-n2"
            icon={<FontAwesomeIcon className="mr-2" icon={faExclamationTriangle} />}
          >
            <h6>
              {intl.formatMessage(
                messages['account.settings.delete.account.modal.text.1'],
                { siteName: getConfig().SITE_NAME },
              )}
            </h6>
            <p>
              {intl.formatMessage(
                messages[deleteAccountModalText2MessageKey],
                { siteName: getConfig().SITE_NAME },
              )}
            </p>
            <p>
              <PrintingInstructions />
            </p>
          </Alert>
          <Form.Group
            for={passwordFieldId}
            isInvalid={errorType !== null}
          >
            <Form.Label className="d-block" htmlFor={passwordFieldId}>
              {intl.formatMessage(messages['account.settings.delete.account.modal.enter.password'])}
            </Form.Label>
            <Form.Control
              name="password"
              id={passwordFieldId}
              type="password"
              value={password}
              onChange={onChange}
            />
            {errorType !== null && (
              <Form.Control.Feedback type="invalid" feedback-for={passwordFieldId}>
                {intl.formatMessage(invalidMessage)}
              </Form.Control.Feedback>
            )}
          </Form.Group>
        </div>

      </AlertModal>
    );
  }
}

ConfirmationModal.propTypes = {
  status: PropTypes.oneOf(['confirming', 'pending', 'deleted', 'failed']),
  errorType: PropTypes.oneOf(['empty-password', 'server']),
  intl: intlShape.isRequired,
  onCancel: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  password: PropTypes.string.isRequired,
};

ConfirmationModal.defaultProps = {
  status: null,
  errorType: null,
};

export default injectIntl(ConfirmationModal);
