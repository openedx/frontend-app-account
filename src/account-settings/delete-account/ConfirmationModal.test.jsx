import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider, injectIntl } from '@edx/frontend-platform/i18n';

// Modal creates a portal.  Overriding createPortal allows portals to be tested in jest.
jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: jest.fn(node => node), // Mock portal behavior
}));

import { ConfirmationModal } from './ConfirmationModal'; // eslint-disable-line import/first

const IntlConfirmationModal = injectIntl(ConfirmationModal);

describe('ConfirmationModal', () => {
  let props = {};

  beforeEach(() => {
    props = {
      onCancel: jest.fn(),
      onChange: jest.fn(),
      onSubmit: jest.fn(),
      status: null,
      errorType: null,
      password: 'fluffy bunnies',
    };
  });

  it('should match default closed confirmation modal snapshot', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <IntlConfirmationModal
            {...props}
          />
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match open confirmation modal snapshot', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <IntlConfirmationModal
            {...props}
            status="pending" // This will cause 'modal-backdrop' and 'show' to appear on the modal as CSS classes.
          />
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match empty password confirmation modal snapshot', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <IntlConfirmationModal
            {...props}
            errorType="empty-password"
            status="pending" // This will cause 'modal-backdrop' and 'show' to appear on the modal as CSS classes.
          />
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
