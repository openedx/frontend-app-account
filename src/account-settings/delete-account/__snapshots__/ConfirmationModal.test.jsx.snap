// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationModal should match default closed confirmation modal snapshot 1`] = `null`;

exports[`ConfirmationModal should match empty password confirmation modal snapshot 1`] = `
[
  <div
    data-focus-guard={true}
    style={
      {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
  <div
    className="pgn__modal-layer"
    data-focus-lock-disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onScrollCapture={[Function]}
    onTouchMoveCapture={[Function]}
    onTouchStart={[Function]}
    onWheelCapture={[Function]}
  >
    <div
      className="pgn__modal-content-container"
    >
      <div
        className="pgn__modal-backdrop"
        data-testid="modal-backdrop"
        onClick={[MockFunction]}
        onKeyDown={[MockFunction]}
      />
      <div
        aria-label="Are you sure?"
        className="pgn__modal pgn__modal-md pgn__modal-default pgn__alert-modal"
        role="dialog"
      >
        <div
          className="pgn__modal-header"
        >
          <h2
            className="pgn__modal-title"
          >
            Are you sure?
          </h2>
        </div>
        <div
          className="pgn__modal-body pgn__modal-body-scroll-top pgn__modal-body-scroll-bottom"
        >
          <div />
          <div
            className="pgn__modal-body-content"
          >
            <div
              className="p-3"
            >
              <div
                className="alert d-flex align-items-start alert-danger mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-circle-exclamation mr-2"
                    data-icon="circle-exclamation"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={{}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zm0-384c13.3 0 24 10.7 24 24l0 112c0 13.3-10.7 24-24 24s-24-10.7-24-24l0-112c0-13.3 10.7-24 24-24zM224 352a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
                      fill="currentColor"
                      style={{}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    A password is required
                  </h6>
                  <p
                    className="text-danger"
                  >
                    Sorry, there was an error trying to process your request. Please try again later.
                  </p>
                </div>
              </div>
              <div
                className="alert d-flex align-items-start alert-warning mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-triangle-exclamation mr-2"
                    data-icon="triangle-exclamation"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={{}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 32c14.2 0 27.3 7.5 34.5 19.8l216 368c7.3 12.4 7.3 27.7 .2 40.1S486.3 480 472 480L40 480c-14.3 0-27.6-7.7-34.7-20.1s-7-27.8 .2-40.1l216-368C228.7 39.5 241.8 32 256 32zm0 128c-13.3 0-24 10.7-24 24l0 112c0 13.3 10.7 24 24 24s24-10.7 24-24l0-112c0-13.3-10.7-24-24-24zm32 224a32 32 0 1 0 -64 0 32 32 0 1 0 64 0z"
                      fill="currentColor"
                      style={{}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    You have selected "Delete My Account". Deletion of your account and personal data is permanent and cannot be undone. localhost will not be able to recover your account or the data that is deleted.
                  </h6>
                  <p>
                    If you proceed, you will be unable to use this account to take courses on localhost.
                  </p>
                  <p>
                    You may also lose access to verified certificates and other program credentials. You can make a copy of these for your records before proceeding with deletion.
                  </p>
                </div>
              </div>
              <div
                className="form-group"
                data-testid="validation-form-group"
              >
                <label
                  className="d-block"
                  htmlFor="passwordFieldId"
                >
                  If you still wish to continue and delete your account, please enter your account password:
                </label>
                <input
                  aria-describedby="passwordFieldId-invalid-feedback"
                  className="form-control is-invalid"
                  id="passwordFieldId"
                  name="password"
                  onChange={[MockFunction]}
                  type="password"
                  value="fluffy bunnies"
                />
                <strong
                  className="invalid-feedback"
                  id="passwordFieldId-invalid-feedback"
                >
                  A password is required
                </strong>
              </div>
            </div>
          </div>
          <div />
        </div>
        <div
          className="pgn__modal-footer"
        >
          <div
            className="pgn__action-row"
          >
            <button
              className="btn btn-link"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Cancel
            </button>
            <button
              className="btn btn-danger"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Yes, Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    data-focus-guard={true}
    style={
      {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
]
`;

exports[`ConfirmationModal should match open confirmation modal snapshot 1`] = `
[
  <div
    data-focus-guard={true}
    style={
      {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />,
  <div
    className="pgn__modal-layer"
    data-focus-lock-disabled="disabled"
    onBlur={[Function]}
    onFocus={[Function]}
    onScrollCapture={[Function]}
    onTouchMoveCapture={[Function]}
    onWheelCapture={[Function]}
  >
    <div
      className="pgn__modal-content-container"
    >
      <div
        className="pgn__modal-backdrop"
        data-testid="modal-backdrop"
        onClick={[MockFunction]}
        onKeyDown={[MockFunction]}
      />
      <div
        aria-label="Are you sure?"
        className="pgn__modal pgn__modal-md pgn__modal-default pgn__alert-modal"
        role="dialog"
      >
        <div
          className="pgn__modal-header"
        >
          <h2
            className="pgn__modal-title"
          >
            Are you sure?
          </h2>
        </div>
        <div
          className="pgn__modal-body pgn__modal-body-scroll-top pgn__modal-body-scroll-bottom"
        >
          <div />
          <div
            className="pgn__modal-body-content"
          >
            <div
              className="p-3"
            >
              <div
                className="alert d-flex align-items-start alert-warning mt-n2"
              >
                <div>
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-triangle-exclamation mr-2"
                    data-icon="triangle-exclamation"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={{}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 32c14.2 0 27.3 7.5 34.5 19.8l216 368c7.3 12.4 7.3 27.7 .2 40.1S486.3 480 472 480L40 480c-14.3 0-27.6-7.7-34.7-20.1s-7-27.8 .2-40.1l216-368C228.7 39.5 241.8 32 256 32zm0 128c-13.3 0-24 10.7-24 24l0 112c0 13.3 10.7 24 24 24s24-10.7 24-24l0-112c0-13.3-10.7-24-24-24zm32 224a32 32 0 1 0 -64 0 32 32 0 1 0 64 0z"
                      fill="currentColor"
                      style={{}}
                    />
                  </svg>
                </div>
                <div>
                  <h6>
                    You have selected "Delete My Account". Deletion of your account and personal data is permanent and cannot be undone. localhost will not be able to recover your account or the data that is deleted.
                  </h6>
                  <p>
                    If you proceed, you will be unable to use this account to take courses on localhost.
                  </p>
                  <p>
                    You may also lose access to verified certificates and other program credentials. You can make a copy of these for your records before proceeding with deletion.
                  </p>
                </div>
              </div>
              <div
                className="form-group"
                data-testid="validation-form-group"
              >
                <label
                  className="d-block"
                  htmlFor="passwordFieldId"
                >
                  If you still wish to continue and delete your account, please enter your account password:
                </label>
                <input
                  aria-describedby=""
                  className="form-control"
                  id="passwordFieldId"
                  name="password"
                  onChange={[MockFunction]}
                  type="password"
                  value="fluffy bunnies"
                />
                <strong
                  className="invalid-feedback"
                  id="passwordFieldId-invalid-feedback"
                >
                  Unable to delete account
                </strong>
              </div>
            </div>
          </div>
          <div />
        </div>
        <div
          className="pgn__modal-footer"
        >
          <div
            className="pgn__action-row"
          >
            <button
              className="btn btn-link"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Cancel
            </button>
            <button
              className="btn btn-danger"
              disabled={false}
              onClick={[MockFunction]}
              type="button"
            >
              Yes, Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    data-focus-guard={true}
    style={
      {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />,
]
`;
